// @domicile/core/dist/orchestrator-sample.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sampleOrchestratorRun = sampleOrchestratorRun;
const manifest_builder_1 = require("./manifest-builder");
const SAMPLE_CONTRACTS = [
    {
        enhancement_area: "Instagram Product Launch",
        objective: "Launch new coffee blend targeting young professionals",
        implementation_plan: {
            modules: ["AudienceAnalyzer", "ContentGenerator", "ChannelPlanner"],
            architecture: "Event-driven services",
            estimated_effort: "3 weeks",
        },
        depends_on: ["Brand Voice Alignment"],
        sources: ["Instagram Insights", "Customer CRM"],
        governance: {
            security: "PII scrubbed at ingestion",
            compliance: "FTC social media guidelines",
            ethics: "Disclose sponsored content",
        },
        validation_criteria: "Engagement lift > 25%",
        confidence_score: 0.86,
    },
];
async function sampleOrchestratorRun() {
    const manifest = (0, manifest_builder_1.buildManifest)(SAMPLE_CONTRACTS);
    console.log("📦 Sample manifest generated:", JSON.stringify(manifest, null, 2));
    return manifest;
}
